sympy/core/basic.py

```diff
@@ -122,6 +122,22 @@ def __new__(cls, *args):
         obj = object.__new__(cls)
         obj._assumptions = cls.default_assumptions
         obj._mhash = None  # will be set by __hash__ method.
+        if not all(isinstance(arg, (Basic, ManagedProperties)) for arg in args ):
+            # XXX: ManagedProperties should be remove eventually,
+            # but this is currently (SymPy 1.9) not possible due to
+            # how Functions work.
```

Comment:
I can't remember any more. What are the cases where a Function is stored in the args? Are any of them in a core part of the codebase?

Reply:
```
sympy/core/tests/test_function.py", line 736, in test_subs_in_derivative
    assert Subs(Derivative(f(f(x)), x), f, cos).doit() == s
```
This mostly.

Reply:
Actually,
`Derivative(f(f(x)), x).subs(f, cos).doit()` seems to work just fine, so I wonder if this is actually really needed. It would be a regression though, but we are deprecating this anyways.

Reply:
What exactly would be a regression?

Reply:
Using `.subs` isn't the same as using `Subs`.

Reply:
> What exactly would be a regression?

No longer being able to use function classes (e.g. `f` and `cos`) as arguments to `Subs`.

> Using `.subs` isn't the same as using `Subs`.

From what I understand, `Subs` is used to delay a substitution, or create a symbolic substitution if you like, because it might change the "meaning" of a derivative, but I don't think that is the case here, both `Subs` and `subs` give the same result. Are there situations where this is not the case (when substituting function classes)?

Reply:
> Are there situations where this is not the case (when substituting function classes)?

The main difference is that `Subs` creates a `Basic` instance holding the objects to be substituted (e.g. the `Function` object) whereas `.subs` can simply create the new expression resulting from substitution directly.

Reply:
I understand that their computational implementation is different, but are there situations where they result in a mathematical different result? Why do you want to be able to use `Subs` with `f` or `cos`? For variables it potentially matters if we substitute them before or after taking the derivative, but are there situations where that matters for functions?

Reply:
> are there situations where that matters for functions?

I'm not sure if this is really well defined at all but maybe something weird like this:
```python
In [6]: f(f(x)).diff(f(x)).subs(f, cos)
Out[6]: 
⎛ d          ⎞│         
⎜───(cos(ξ₀))⎟│         
⎝dξ₀         ⎠│ξ₀=cos(x)
```

Reply:
I think
```python
Subs(Derivative(f(f(x)), f(x)), f, cos).doit()
Derivative(f(f(x)), f(x)).subs(f, cos).doit()
f(f(x)).diff(f(x)).subs(f, cos).doit()
```
Are all the same

Reply:
For variables, it matters a lot:
```python
>>> Derivative(sin(x),x).subs(x,0).doit()
1
>>> Derivative(sin(x).subs(x,0),x).doit()
0
```
whether we sub first or take the derivative first.

Reply:
Maybe something like this:
```python
In [6]: cos(x).diff(x, 2).subs(cos, exp)
Out[6]: 
  x
-ℯ 

In [7]: cos(x).subs(cos, exp).diff(x, 2)
Out[7]: 
 x
ℯ 
```
I'm not really sure what kind of example we're looking for here.

Reply:
Interesting, it seems that we can't diff w.r.t. `cos(x)` but we can do it w.r.t. `f(x)` (due to how `_diff_wrt` is set).

Reply:
Ah I see, yes, if you are substituting two different defined functions then the order is going to matter. That's unfortunate.

Reply:
Is that something that's tested? There's not a lot of clarity on the difference between `subs` and `replace`, but I would generally expect something like that to only be supported by `replace`. 

It would be possible to that symbolically using something more explicit like `Subs(exp(x), exp(a), cos(a))`, where `a` is a `Wild`. But I rather expect this is something that is only supported accidentally.

Reply:
I feel that I've not really stated my concerns clearly on this topic:

`Subs` is a SymPy object which is created when `expr.subs` is used if expr contains a `Derivative` object. This is because someone might want to represent symbolically the derivative `f'(x)` at `x=0`. Here the substitution then also needs to be represented symbolically because `f '(0)` would be equal to zero since it is constant.

`.subs` has been specifically coded to work with substitutions of functions (which are classes) as in `f(x).subs(f, cos) == cos(x)`.

When subs is used for function substitutions and there is a `Derivative` in that expression, a `Subs` object is created that now has as its arguments `Function`s, which are subclasses of `Basic` and not instances of `Basic`. 

I was hoping the creation of such a `Subs` object could be avoided altogether by simply performing the actual substitution rather than creating the symbolic `Subs` object because I thought there would never be an ambiguity (unlike in the case of substituting a variable with a constant), but it seems that this is not the case.

A unit test exists for a `Subs` object with `Function`s as its arguments: sympy/core/tests/test_function.py, line 736

Reply:
Ultimately the problem is that functions aren't instances of Basic. In the mean time maybe it's possible for `Subs` to store a function using a dummy so it actually stores the called function like `f(_x)`.

Reply:
I considered something like this too, but it would leave some problems:
- how do we distinguish between a substitution with a function and substitution with a function as a function of a dummy? Do we store the dummy in the `Subs` object too if we create it for this purpose?
- The printing methods would have to be rewritten to take this into account.

Reply:
Yes, it would leave some problems. If a function was a proper symbolic object then I doin't think that Subs would be needed at all. We use it for something like this:
```python
In [1]: f(x).diff(x).subs(x, 0)
Out[1]: 
⎛d       ⎞│   
⎜──(f(x))⎟│   
⎝dx      ⎠│x=0
```
If Function was a proper symbolic object though then we could represent that as just `Derivative(f)(0)` which would in turn print as `f'(0)`. Then there's no need for Subs and no need to specify any symbols at all.

Reply:
We can't really "allow" Basic subclasses because they generally won't work:
```python
>>> from sympy import *
>>> f = Function('f')
>>> x = Symbol('x')
>>> e = Subs(f(x), f, cos)
```
If you check `dir(Expr)` it's clear that a large number of methods don't work and it's not reasonable to make them work e.g.:
```python
>>> e.apart()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/oscar/current/sympy/sympy.git/sympy/core/expr.py", line 3696, in apart
    return apart(self, x, **args)
  File "/home/oscar/current/sympy/sympy.git/sympy/utilities/decorator.py", line 35, in threaded_func
    return func(expr, *args, **kwargs)
  File "/home/oscar/current/sympy/sympy.git/sympy/polys/partfrac.py", line 81, in apart
    (P, Q), opt = parallel_poly_from_expr((P, Q), x, **options)
  File "/home/oscar/current/sympy/sympy.git/sympy/polys/polytools.py", line 4403, in parallel_poly_from_expr
    return _parallel_poly_from_expr(exprs, opt)
  File "/home/oscar/current/sympy/sympy.git/sympy/polys/polytools.py", line 4440, in _parallel_poly_from_expr
    expr = expr.expand()
  File "/home/oscar/current/sympy/sympy.git/sympy/core/cache.py", line 70, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/oscar/current/sympy/sympy.git/sympy/core/expr.py", line 3627, in expand
    expr, hit = Expr._expand_hint(expr, hint, deep=deep, **hints)
  File "/home/oscar/current/sympy/sympy.git/sympy/core/expr.py", line 3560, in _expand_hint
    arg, arghit = Expr._expand_hint(arg, hint, **hints)
  File "/home/oscar/current/sympy/sympy.git/sympy/core/expr.py", line 3560, in _expand_hint
    arg, arghit = Expr._expand_hint(arg, hint, **hints)
  File "/home/oscar/current/sympy/sympy.git/sympy/core/expr.py", line 3559, in _expand_hint
    for arg in expr.args:
TypeError: 'property' object is not iterable
```
That's because `f.args` gives a property rather than a tuple which breaks a fundamental invariant that is supposed to be expected from Basic.